### PR TITLE
extend boot phase information into PCR 11 during boot

### PR DESCRIPTION
**Pull request by poettering**
Still work in progress. Lack docs and stuff. Don't merge yet.

### CODE REVIEW

File: src/boot/pcrphase.c [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+static int help(int argc, char *argv[], void *userdata) {
+        _cleanup_free_ char *link = NULL;
+        int r;
+
+        r = terminal_urlify_man("systemd-pcrphase", "1", &link);
+        if (r < 0)
+                return log_oom();
+
+        printf("%1$s  [OPTIONS...] COMMAND ...\n"
+               "\n%5$sMeasure boot phase into TPM2 PCR 11.%6$s\n"
```
Comment (bluca) [inline src/boot/pcrphase.c:30]: shouldn't the PCR be selectable with a switch (defaulting to 11)?

Comment (poettering) [reply src/boot/pcrphase.c:30]: so far all our PCR assignments are hardcoded, and defined as macros in `src/fundamental/tpm-pcr.h`. We use exactly those definitions here, too.

I really want to push people to use a common vocabulary of TPM assignments, much like the kernel also doesn't make it configurable to use any other PCR than 9 for the initrds. 

Should this ever become an issue we can certainly revisit this: making something that is currently hardcoded configurable is always easy.

**Comment by poettering** [comment]
A companion to this is https://github.com/systemd/mkosi/pull/1194 which ensures the PCR extension takes places in the initrd as intended.

**Comment by poettering** [comment]
Has all docs now. Ready for review.

Like with the other EFI stub stuff, no tests. But once we have https://github.com/systemd/mkosi/pull/1194 and https://github.com/systemd/mkosi/pull/1193 merged we might just use mkosi tests without much additional magic for that. Will work on this separately once the mkosi side and this is merged.

**Comment by bluca** [comment]
probably related?

```
17:42:19 FAILED TASKS:
17:42:19 -------------
17:42:19 TEST-70-TPM2
```

**Comment by bluca** [comment]
```
17:30:11 [   21.989147] testsuite-70.sh[475]: + cmp - /tmp/result
17:30:11 [   21.995942] testsuite-70.sh[474]: Measuring boot phases: enter-initrd, enter-initrd:leave-initrd, enter-initrd:leave-initrd:ready
17:30:11 [   21.996327] testsuite-70.sh[474]: # PCR[11] Phase <enter-initrd>
17:30:11 [   21.996984] testsuite-70.sh[474]: # PCR[11] Phase <enter-initrd:leave-initrd>
17:30:11 [   21.997751] testsuite-70.sh[474]: # PCR[11] Phase <enter-initrd:leave-initrd:ready>
17:30:11 [   21.998532] testsuite-70.sh[475]: - /tmp/result differ: byte 9, line 1
```

File: src/boot/measure.c [hunk 2]
```diff
@@ -87,6 +90,21 @@ static int help(int argc, char *argv[], void *userdata) {
         return 0;
 }
 
+static char *normalize_phase(const char *s) {
```
Comment (bluca) [inline src/boot/measure.c:95]: s is unused?

File: src/boot/measure.c [hunk 3]
```diff
@@ -87,6 +90,21 @@ static int help(int argc, char *argv[], void *userdata) {
...
+        _cleanup_strv_free_ char **l = NULL;
+
+        /* Let's normalize phase expressions. We split the series of colon-separated words up, then remove
+         * all empty ones, and glue them back together again. In other words we remove duplicate ":", as well
+         * as leading and trailing ones. */
+
+        l = strv_split(optarg, ":"); /* Split series of words */
```
Comment (bluca) [inline src/boot/measure.c]: did you mean `s` here?

File: src/boot/measure.c (hunk 2 above)
Comment (poettering) [reply src/boot/measure.c:95]: ah, fun. I guess this only worked, since the only param we call this with is optarg anyway...

**Comment by poettering** [comment]
Force pushed a new version, which hopefully passes CI, and fixes the issues pointed out. PTAL

**Comment by bluca** [comment]
doesn't look related

```
16:50:56 ======================================================================
16:50:56 FAIL: test_udev_link_file (__main__.NetworkctlTests)
16:50:56 ----------------------------------------------------------------------
16:50:56 Traceback (most recent call last):
16:50:56   File "./test/test-network/systemd-networkd-tests.py", line 1050, in test_udev_link_file
16:50:56     self.assertRegex(output, r'Link File: -')
16:50:56 AssertionError: Regex didn't match: 'Link File: -' not found in '● 1: lo                         \n                     Link File: n/a\n                  Network File: n/a\n                          Type: loopback\n                          Kind: n/a\n                         State: carrier (unmanaged)\n                  Online state: unknown\n              Hardware Address: 00:00:00:00:00:00\n                           MTU: 65536\n                         QDisc: noqueue\n  IPv6 Address Generation Mode: eui64\n      Number of Queues (Tx/Rx): 1/1\n                       Address: 127.0.0.1\n                                ::1\n             Activation Policy: up\n           Required For Online: yes'
16:50:56 
16:50:56 ----------------------------------------------------------------------
16:50:56 Ran 155 tests in 1388.951s
16:50:56 
16:50:56 FAILED (failures=1, skipped=1, expected failures=10)
```

**Review by bluca — APPROVED** [review]

**Pull request merged**